### PR TITLE
Add instance redirects!

### LINKED ISSUE #1873: [Feature request] Add UI button redirect to a healthier instance when watching a video on a particular instance
**Is your feature request related to a problem? 

It'd be great if there were a button or drop down menu to select another public instance.

**Describe the solution you'd like**

For example,

For example, I would like to have a button on the interface for the video link below so I can choose a different instance if the snoypta instance is not cutting it that day for whatever reason:

https://invidious.snopyta.org/watch?v=2s9fM0CDAQ4

### PR DESCRIPTION
**Pull request by syeopite**
Add functions that redirect users to `redirect.invidious.io` or alternatively, if they choose so, redirect to a random update-to-date and healthy clear-net instance.

There's also a few front end additions.

The jet icon here is just a placeholder. I couldn't find any v4 ion icon that fits.

There's also a search redirect for when searching is broken. The filtering UI has been hidden to provide prominence to that message.
 redirect-3

The error template has also been enhanced to contain a link to Youtube, to switch Invidious instance, and to refresh the page.

error msg

Anyhow, this feature works by fetching an instance list from [api.invidious.io](https://api.invidious.io) so the load on that server would increase if merged. 

Closes #1873
Supersedes #1206

### CODE REVIEW

**Comment by syeopite** [comment]
Related #944

**Comment by TheFrenchGhosty** [comment]
>Anyhow, this feature works by fetching an instance list from api.invidious.io so the load on that server would increase if merged.

This isn't really a problem, the server has enough bandwidth to handle the requests, so this is fine.

File: src/invidious/routes/misc.cr [hunk 1]
```diff
@@ -35,4 +35,15 @@ class Invidious::Routes::Misc < Invidious::Routes::BaseRoute
     locale = LOCALES[env.get("preferences").as(Preferences).locale]?
     rendered "licenses"
   end
+
+  def redirect(env)
```
Comment (saltycrys) [inline src/invidious/routes/misc.cr:39]: Please change this to be generic, something like https://github.com/iv-org/invidious/blob/c9a316ad353cd1ee38c34d1aff0759765b37b2b3/src/invidious/views/template.ecr#L63

**Review by saltycrys — CHANGES_REQUESTED** [review]

**Comment by unixfox** [comment]
Is this really a good idea? What if it redirects to an instance that is down? api.invidious.io doesn't do a health check so that can happen.

Wouldn't it be better to redirect to https://redirect.invidious.io instead? This way the user can choose whatever instance he wants instead of random one and has a clear list of all the existing public instances.

**Comment by TheFrenchGhosty** [comment]
Mmh I kinda agree with @unixfox here.

It would also simplify the code a lot.

**Comment by syeopite** [comment]
> Is this really a good idea? What if it redirects to an instance that is down? api.invidious.io doesn't do a health check so that can happen.

Good point though the chances of running into an dead instance are pretty slim. I instance hop frequently and I've only ever encountered an instance that is down once. It's far more likely to reach a outdated instance instead. The latter of which can be solved if we just compared the versions to the current one.

Personally I prefer this version more than the only redirect to [https://redirect.invidious.io](https://redirect.invidious.io). Its less work on the users end and even allows for site routes to pass through for JS disabled users.

**Review by saltycrys — APPROVED** [review]

**Comment by Perflyst** [comment]
How do we continue with this one? I also prefer redirect.invidious.io...

> Personally I prefer this version more than the only redirect to https://redirect.invidious.io. Its less work on the users end and even allows for site routes to pass through for JS disabled users.

redirect.invidious.io works without javascript too (not completely though)

**Comment by syeopite** [comment]
Compromise and allow it to be configurable?

**Comment by syeopite** [comment]
There we go. All of these concerns has been addressed.

**Comment by unixfox** [comment]
What happens if api.invidious.io is down? Does it skip the rendering of the link?

**Comment by syeopite** [comment]
`redirect.invidious.io` would be used instead.

**Comment by syeopite** [comment]
I'll like to add to my previous argument about potentially redirecting to instances that are down. `api.invidious.io` currently refreshes its instance list every 5 minutes; any instance that is down for longer than that would quickly be removed from the API as well.

**Comment by unixfox** [comment]
> I'll like to add to my previous argument about potentially redirecting to instances that are down. `api.invidious.io` currently refreshes its instance list every 5 minutes; any instance that is down for longer than that would quickly be removed from the API as well.  

Not if the API starts to freeze, which often happens and need a restart.

**Comment by TheFrenchGhosty** [comment]
@unixfox The api has a cronjob that restart it daily

**Comment by unixfox** [comment]
> @unixfox The api has a cronjob that restart it daily

If the API happens to freeze before the restart for whatever reason, this system will still offer instances that could be down until the API is restarted.

Unless this system rely on a reliable healthcheck service like uptime robot, there will be cases where an instance that is down could be offered to the user.

**Comment by syeopite** [comment]
Even if one got redirected to an instance that is down —  the chance of which is already extremely low — how is it any different from clicking on an instance that is down on redirect.invidious.io? In both cases you'll have to go back to the original page and pick a new instance or in the case of this PR click on the switch instance button again

**Comment by unixfox** [comment]
> Even if one got redirected to an instance that is down — the chance of which is already extremely low — how is it any different from clicking on an instance that is down on redirect.invidious.io? In both cases you'll have to go back to the original page and pick a new instance or in the case of this PR click on the switch instance button again

On redirect.invidious.io, multiple instances get presented to you. With your system, only one is given.
If an instance is down from redirect.invidious.io then the user will probably understand a bit better that he can try another one because he saw multiple instances available.
Whereas with your system just one link is given and the user may not understand directly that it can come back to the previous page THEN refresh the page for getting another instance.

**Comment by syeopite** [comment]
Good point, though you exactly don't need to refresh the page to head to another instance. Just going back to the previous page would be enough to get reconnected to another instance as any connection to the `/redirect` endpoint would cause a redirect.

**Pull request merged**